data-structures.md

```diff
@@ -313,11 +305,15 @@ if ((shift <size) && (sign bit of byte is set))
 
 Datastructures in Filecoin are encoded as compactly as is reasonable. At a high level, each object is converted into an ordered array of its fields (ordered by their appearance in the struct declaration), then CBOR marshaled, and prepended with an object type tag.
```

Comment:
I had a chat with @pooja last week, and the optimization goals of our encoding-mix came up again and it still remains unclear (to me at least). Not strictly part of this changeset, but this sentence sums it up very well for me:

> each object is converted into an ordered array of its fields (ordered by their appearance in the struct declaration), then CBOR marshaled, and prepended with an object type tag.

We drop the bigger context of the named fields (that CBOR would support), convert it into an unnamed mixed-data array, which then is CBOR'ed adding the meta-information that it is an array as well as the type-information for each field - both unnecessary if we need to hard know the typed fields of the struct anyways. Leading to a weird, seemingly inconsistent mix of sometimes optimizing away type info in the name of "saving data", but sometimes keeping it with the "it's still good to have some context". The same for FCS, which contains some typed info, yet it is not possible to decode without knowing the wider context and struct. I am not sure what we are optimizing for here, not for saving space or we'd drop this - IMHO useless - type info...

Reply:
The type information in CBOR doesnt actually take up much (if any) extra space. For most fields, the type and field length are combined into a single byte. Yes, we could maybe save a couple bytes by switching to a fully custom format, but a big part of the point here is that its not a fully custom format. It's CBOR, you can inspect the fields without necessarily requiring filecoin specific code or context, you can traverse the links with standard IPLD tooling, you can use off-the-shelf tooling for all of this. Yes, you can't convert the bytes of an address to its correct string representation without filecoin specific code, but that doesnt invalidate the usefulness of having the majority of the codec already written for you. 

I'm not entirely against a fully custom codec, but I just don't think its that worth it.